File: web/src/api/futures.ts
import { useQuery } from '@tanstack/react-query';
import client from './client';

export interface FuturesContractOut {
  code: string;
  name: string;
  exchange: string;
  exchange_label: string | null;
  product: string;
  underlying_instrument: string | null;
  contract_size: string | null;
  price_unit: string | null;
  quote_unit: string | null;
  is_main: boolean;
  list_date: string | null;
  delist_date: string | null;
  last_seen_at: string | null;
}

export interface FuturesContractListResponse {
  items: FuturesContractOut[];
  total: number;
  page: number;
  page_size: number;
}

export interface FuturesDashboardSection {
  product: string;
  product_label: string | null;
  items: FuturesDailyBarOut[];
  best_performer: FuturesDailyBarOut | null;
  worst_performer: FuturesDailyBarOut | null;
  count: number;
}

export interface FuturesDashboardResponse {
  sections: FuturesDashboardSection[];
  trade_date: string | null;
  total_contracts: number;
}

export interface FuturesDailyBarOut {
  code: string;
  trade_date: string;
  open: string | null;
  high: string | null;
  low: string | null;
  close: string | null;
  settle: string | null;
  pre_settle: string | null;
  volume: number | null;
  open_interest: number | null;
  turnover: string | null;
  warehouse_receipts: number | null;
  settle_change_pct: number | null;
  change_pct: number | null;
}

export interface FuturesLeaderboardRow {
  code: string;
  name: string;
  exchange: string;
  product: string;
  close: string | null;
  settle: string | null;
  pre_settle: string | null;
  change_pct: number | null;
  volume: number | null;
  open_interest: number | null;
  turnover: string | null;
}

export interface FuturesLeaderboardResponse {
  items: FuturesLeaderboardRow[];
  direction: 'gainers' | 'losers';
  exchange: string | null;
  trade_date: string | null;
}

export interface FuturesStats {
  total_contracts: number;
  total_bars: number;
  latest_trade_date: string | null;
}

const BASE = '/futures';

export const futuresApi = {
  contracts: (params?: { exchange?: string; product?: string; is_main?: boolean; search?: string; page?: number; page_size?: number }) => {
    const parts: string[] = [];
    const append = (k: string, v: unknown) => {
      if (v === undefined || v === null || v === '') return;
      parts.push(`${encodeURIComponent(k)}=${encodeURIComponent(String(v))}`);
    };
    append('exchange', params?.exchange);
    append('product', params?.product);
    if (params?.is_main !== undefined) append('is_main', params.is_main);
    append('search', params?.search);
    append('page', params?.page ?? 1);
    append('page_size', params?.page_size ?? 200);
    const qs = parts.length === 0 ? '' : `?${parts.join('&')}`;
    return client.get<FuturesContractListResponse>(`${BASE}/contracts${qs}`);
  },
  dashboard: () => client.get<FuturesDashboardResponse>(`${BASE}/dashboard`),
  leaderboard: (direction: 'gainers' | 'losers', exchange?: string, top = 30) => {
    const parts = [`direction=${direction}`, `top=${top}`];
    if (exchange) parts.push(`exchange=${encodeURIComponent(exchange)}`);
    return client.get<FuturesLeaderboardResponse>(`${BASE}/leaderboard?${parts.join('&')}`);
  },
  stats: () => client.get<FuturesStats>(`${BASE}/stats`),
};

export function useFuturesDashboard() {
  return useQuery({
    queryKey: ['futures', 'dashboard'],
    queryFn: async () => {
      const res = await futuresApi.dashboard();
      return res.data;
    },
    staleTime: 60_000,
  });
}

export function useFuturesLeaderboard(direction: 'gainers' | 'losers', exchange?: string) {
  return useQuery({
    queryKey: ['futures', 'leaderboard', direction, exchange ?? null],
    queryFn: async () => {
      const res = await futuresApi.leaderboard(direction, exchange);
      return res.data;
    },
    staleTime: 60_000,
  });
}

export function useFuturesStats() {
  return useQuery({
    queryKey: ['futures', 'stats'],
    queryFn: async () => {
      const res = await futuresApi.stats();
      return res.data;
    },
    staleTime: 60_000,
  });
}

File: app/schemas/futures.py
"""Pydantic schemas for the futures API surface.

Covers contract metadata, daily OHLCV bars, dashboard aggregations,
and exchange leaderboards.
"""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Contract metadata
# ---------------------------------------------------------------------------


class FuturesContractOut(BaseModel):
    """A row in the futures contract list."""

    code: str = Field(..., description="Main contract code, e.g. CU0")
    name: str = Field(..., description="Display name")
    exchange: str = Field(..., description="Exchange code: SHFE/DCE/CZCE/CFFEX/INE/GFEX")
    exchange_label: str | None = Field(None, description="Chinese exchange name")
    product: str = Field(..., description="Category: 金属/能源化工/农产品/金融期货")
    underlying_instrument: str | None = Field(None, description="Current leading specific contract code")
    contract_size: Decimal | None = None
    price_unit: str | None = None
    quote_unit: str | None = None
    is_main: bool = True
    list_date: date | None = None
    delist_date: date | None = None
    last_seen_at: datetime | None = None


class FuturesContractListResponse(BaseModel):
    """Paginated futures contract list response."""

    items: list[FuturesContractOut]
    total: int
    page: int = 1
    page_size: int = 200


# ---------------------------------------------------------------------------
# Daily bars
# ---------------------------------------------------------------------------


class FuturesDailyBarOut(BaseModel):
    """A single daily OHLCV row for a futures contract."""

    code: str
    trade_date: date
    open: Decimal | None = None
    high: Decimal | None = None
    low: Decimal | None = None
    close: Decimal | None = None
    settle: Decimal | None = None
    pre_settle: Decimal | None = None
    volume: int | None = None
    open_interest: int | None = None
    turnover: Decimal | None = None
    warehouse_receipts: int | None = None

    # Derived: percentage change from pre_settle -> settle
    settle_change_pct: float | None = None
    # Derived: percentage change from pre_close -> close
    change_pct: float | None = None


class FuturesDailyBarListResponse(BaseModel):
    """Response wrapper for daily bars."""

    items: list[FuturesDailyBarOut]
    count: int
    code: str | None = None


# ---------------------------------------------------------------------------
# Dashboard
# ---------------------------------------------------------------------------


class FuturesDashboardSection(BaseModel):
    """Latest-day data for all contracts in one product group."""

    product: str
    product_label: str | None = None
    items: list[FuturesDailyBarOut]
    best_performer: FuturesDailyBarOut | None = None
    worst_performer: FuturesDailyBarOut | None = None
    count: int


class FuturesDashboardResponse(BaseModel):
    """Dashboard data grouped by product category."""

    sections: list[FuturesDashboardSection]
    trade_date: date | None = None
    total_contracts: int


# ---------------------------------------------------------------------------
# Leaderboard
# ---------------------------------------------------------------------------


class FuturesLeaderboardRow(BaseModel):
    """A single row on the leaderboard."""

    code: str
    name: str
    exchange: str
    product: str
    close: Decimal | None = None
    settle: Decimal | None = None
    pre_settle: Decimal | None = None
    change_pct: float | None = None
    volume: int | None = None
    open_interest: int | None = None
    turnover: Decimal | None = None


class FuturesLeaderboardResponse(BaseModel):
    """Sorted leaderboard (gainers / losers)."""

    items: list[FuturesLeaderboardRow]
    direction: str = Field("gainers", description="gainers or losers")
    exchange: str | None = None
    trade_date: date | None = None


# ---------------------------------------------------------------------------
# Service-level filters
# ---------------------------------------------------------------------------


class FuturesFilterParams(BaseModel):
    """Internal filter params for the contract list endpoint."""

    exchange: str | None = None
    product: str | None = None
    is_main: bool | None = True
    search: str | None = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=200, ge=1, le=500)
